File: benchmarks/agents/memory/CircularList.py
import collections


# Class storing a circular index similar to polar coordinate.
CircularIndex = collections.namedtuple("CircularIndex", field_names=["angle", "radius"])


class CircularList:
    """
    A list that can contain an infinite number of elements where:
     - removal of the first element is O(N / C)
     - addition of an element at the end of the list is amortized O(1)
     - access and modification of elements are O(1)

    N = number of elements in the list
    C = the estimated size of the list (how big we expect the list to be)
    """

    def __init__(self, expected_size):
        """
        Create a circular list.
        :param expected_size: the estimated size of the list (how big we expect the list to be)
        """
        self.expected_size = expected_size
        self.elements = [[] for _ in range(expected_size)]
        self.first_index = 0
        self.current_index = 0

    def append(self, element):
        """
        Add an element at the end of the list.
        :param element: the element to add
        """
        c_index = self.circular_index(self.current_index, first_index=0)
        self.elements[c_index.angle].append(element)
        self.current_index += 1

    def pop(self):
        """
        Remove the first element from the list.
        :return: the removed element
        """
        c_index = self.circular_index(self.first_index, first_index=0)
        element = self.elements[c_index.angle][0]
        self.elements[c_index.angle] = self.elements[c_index.angle][1:]
        self.first_index += 1
        return element

    def __getitem__(self, index):
        """
        Retrieve an element from the list.
        :param index: the index of the element to retrieve
        :return: the element
        """
        if index >= len(self):
            return None
        c_index = self.circular_index(index)
        return self.elements[c_index.angle][c_index.radius]

    def __setitem__(self, index, element):
        """
        Replace an element in the list.
        :param index: the index of the element to replace
        :param element: the new element
        """
        c_index = self.circular_index(index)
        self.elements[c_index.angle][c_index.radius] = element

    def circular_index(self, index, first_index=None):
        """
        Compute the circular index corresponding to the provided index.
        :param index: the index whose circular index must be computed
        :param first_index: the index to consider as the first element in the circular list
        :return: the circular index
        """
        if index < 0:
            index += len(self)
        if first_index is None:
            first_index = self.first_index
        return CircularIndex((index + first_index) % self.expected_size, index // self.expected_size)

    def __len__(self):
        """
        Retrieve the number of elements currently stored in the list.
        :return: the number of elements currently stored in the list
        """
        return self.current_index - self.first_index

    def clear(self):
        """
        Clear the content of the circular list.
        """
        self.elements = [[] for _ in range(self.expected_size)]
        self.current_index = 0
        self.first_index = 0

File: benchmarks_cpp/src/circular_list.cpp
#include "circular_list.hpp"

CircularIndex::CircularIndex(int angle, int radius) {
    this->angle = angle;
    this->radius = radius;
}

template<class T>
CircularList<T>::CircularList(int expected_size) {
    this->expected_size = expected_size;
    this->elements = new T *[expected_size]{};
    this->first_index = 0;
    this->current_index = 0;
}

template<class T>
void CircularList<T>::append(T element) {
    CircularIndex c_index = this->circularIndex(this->current_index, 0);
    int n = this->length() / this->expected_size + 1;
    std::cout << "list.append: [n: " << n << "]" << std::endl;
    auto elements = new T [n]{};
    for (auto i = 0; i < n - 1; i++) {
        elements[i] = this->elements[c_index.angle][i];
    }
    elements[n - 1] = element;
    delete this->elements[c_index.angle];
    this->elements[c_index.angle] = std::move(elements);
    this->current_index += 1;
}

template<class T>
T CircularList<T>::pop() {

    // Retrieve the element that must be removed from the list.
    CircularIndex c_index = this->circularIndex(this->first_index, 0);
    T element = this->elements[c_index.angle][0];

    // Remove the element from the list.
    int n = this->length() / this->expected_size + 1;
    auto elements = new T [n]{};
    for (auto i = 0; i < n; i++) {
        elements[i] = this->elements[c_index.angle][i + 1];
    }
    delete this->elements[c_index.angle];
    this->elements[c_index.angle] = elements;

    // Increase the first index, and return the removed element.
    this->first_index += 1;
    return element;
}

template<class T>
T &CircularList<T>::operator[](int index) {
    CircularIndex c_index = this->circularIndex(index);
    return this->elements[c_index.angle][c_index.radius];
}

template<class T>
CircularIndex CircularList<T>::circularIndex(int index, int first_index) {
    if (index < 0) {
        index += this->length();
    }
    if (first_index == -1) {
        first_index = this->first_index;
    }
    return CircularIndex((index + first_index) % this->expected_size, index / this->expected_size);
}

template<class T>
int CircularList<T>::length() {
    return this->current_index - this->first_index;
}

template<class T>
void CircularList<T>::clear() {
    for (auto i = 0; i < this->length(); i++) {
        CircularIndex index = this->circularIndex(i);
        delete this->elements[index.angle];
        this->elements[index.angle] = nullptr;
    }
    delete this->elements;
    this->elements = new T *[expected_size]{};
    this->current_index = 0;
    this->first_index = 0;
}

template<class T>
int CircularList<T>::getCurrentIndex() {
    return this->current_index;
}
